Automattic/wp-calypso
# Guided Tours: fix the `tourBranching` function to be able to descend render props

This patch fixes a regression in the `tourBranching` function that was introduced in #22862. That PR converted `Step` children to a functional render prop. That led to `tourBranching` being unable to descend into the children when examining the tour tree.

This gets fixed by `tourBranching` checking if `children` is a function, calling it when it is, and continuing to traverse the tree using the returned children.

Additional changes:
- rename the `config-parsing.js` file to `tour-branching.js` --- more appropriate name.
- simplify the algorithm how tour branches are collected and converted into objects.
- add a unit test for the `tourBranching` function!

**User-visible behavior**
The tour's ability to skip disabled or irrelevant steps was broken. Consider the following tour:
```jsx
<Tour name="jetpack">
  <Step name="init"><Next step="view-site" /></Step>
  <Step name="view-site" when={ isSitePreviewable }><Next step="themes" /></Step>
  <Step name="themes" />
</Tour>
```
When site is previewable, the `view-site` step is displayed. When it's not, clicking "Next" in `init` should skip all the way to `themes`. This was broken for 5 months since #22862. This patch makes that work again.

**How to test:**
Go to `/?tour=jetpack&flags=-preview-layout`. That should make `isSitePreviewable` return `false` and start the "Jetpack Basic" tour. Some steps should be skipped.

**Before the patch:** Instead of skipping the "View Site" step, the tour goes back to the initial step.
**After the patch:** The "View Site" step is skipped and the "Customize" item is higlighted and described.
